[NFC] Refactor symbol table creation in binary writer

wingo opened the pull request:
This commit refactors the representations of symbols in the binary writer.  It will allow us to fix #1534 in a followup.

wingo commented on src/binary-writer.cc:
```diff
@@ -97,10 +97,97 @@ struct RelocSection {
+  bool isSection() const { return type() == Section::type; }
+  bool isEvent() const { return type() == Event::type; }
+
+  const Function& asFunction() const {
```
Hum, I see that I have SpiderMonkey conventions at my fingers; would need to fix this to AsFunction I guess (and same for IsFunction etc above)

binji replied:
Yep, that would be nice, thanks!

binji commented on src/binary-writer.cc:
```diff
@@ -312,34 +401,43 @@ Index BinaryWriter::GetEventVarDepth(const Var* var) {
+  auto iter = symtab_.find(name);
+  if (iter != symtab_.end()) {
+    Index sym_index = iter->second;
+    assert(symbols_[sym_index].type() == T::type);
```
I'm not sure whether this assert is safe; seems like it could be invalid based on malformed data in wasm module. @sbc100 do you know?

wingo replied:
Good point; let's print a warning in that case instead of failing.

In any case, the current logic doesn't actually follow the toolchain-conventions invariants; symbols that are neither imported nor exported by a module have no name uniqueness requirements AFAIU.  I will fix the logic in a followup.

binji commented on src/binary-writer.cc:
```diff
@@ -312,34 +401,43 @@ Index BinaryWriter::GetEventVarDepth(const Var* var) {
     case RelocType::FuncIndexLEB:
-      name = module_->funcs[index]->name;
-      break;
+      if (index < module_->num_func_imports)
```
nit: {} braces around if

binji commented on src/binary-writer.cc:
```diff
@@ -312,34 +401,43 @@ Index BinaryWriter::GetEventVarDepth(const Var* var) {
-      type = SymbolType::Global;
-      name = module_->globals[index]->name;
-      break;
+      if (index < module_->num_global_imports)
```
same here

binji reviewed (approved):
Looks good, but let @sbc100 take a look too. Can you modify/add on to one of the existing tests (perhaps `test/binary/linking_section.txt`?)

wingo commented:
Updated patch addresses review comments, except regarding tests: because this patch has no functional change, no test needs to be modified.  I'll extend the `linking-section.txt` test in the followup though; thanks for the pointer!

wingo commented:
(And apologies wrt force-pushes; I gather the convention is to push commits to fix nits, then you rebase/squash when landing -- will do that next time.)

sbc100 commented:
> (And apologies wrt force-pushes; I gather the convention is to push commits to fix nits, then you rebase/squash when landing -- will do that next time.)

I don't think we have any rules around this.   I basically always use a rebase flow which means I often/always have to force push to github PRs (even if i don't squash, which I also occasionally do too).

sbc100 commented on src/binary-writer.cc:
```diff
@@ -97,10 +97,97 @@ struct RelocSection {
+  };
+  struct Section {
+    static const SymbolType type = SymbolType::Section;
+    Index section;
```
Why not call this `index` like the others?   Or the other way around and don't call any of them `index`?

wingo replied:
I was using the names from https://github.com/WebAssembly/tool-conventions/blob/master/Linking.md, where this field is called "section".

sbc100 commented on src/binary-writer.cc:
```diff
@@ -97,10 +97,97 @@ struct RelocSection {
+  bool explicit_name() const { return flags() & WABT_SYMBOL_FLAG_EXPLICIT_NAME; }
+  bool no_strip() const { return flags() & WABT_SYMBOL_FLAG_NO_STRIP; }
+
+  bool IsFunction() const { return type() == Function::type; }
```
Why not just use `SymbolType::Function` here?  What purpose does the `Function::type` serve?

wingo replied:
The purpose of the `::type` definitions is to allow `InternSymbol` to know what kind of symbol to make.  I admit that I wasn't sure which name to use here in `IsFunction` and friends; if you prefer that I change it, I am happy to do so.

sbc100 replied:
I guess it will make more sense if/when we add data symbol but for now it looks like you could avoid a lot of complexity and avoid InternSymbol being a template at all by just passing the SymbolType to InternSymbol.

It seems like having a separate class nested inside of `Symbol`, for each different type of symbol payload adds a fair amount of complexity, but perhaps that is justified and will make more sense once the followups land?

lgtm aside from this minor concern.

wingo replied:
I see what you are saying, and if every symbol just had an `Index` payload it would be entirely clear-cut.  I guess since I saw the SymbolType::Data symbols having additional fields, I found it confusing for Symbol to have unused fields for e.g. SymbolType::Function symbols, which would propagate around as arguments to e.g. InternSymbol and related functions.  But that is a theoretical concern given that wabt doesn't know anything about data symbols, and my patches don't change that.

The followup #1537 doesn't do anything amazing here; this is just a question of data representation.  If you like I can change to a less structured `class Symbol { string name; uint8_t flags; SymbolType type; Index index; Offset offset; Address size; }` and assert that offset/size accessors only happen for data symbols (of which we will have none), or even exclude data symbols from the representation.  WDYT?

sbc100 replied:
I think I'm good with this approach.  What do you think binji?

binji replied:
I'm happy either way, tbh. I like having more explicit types (i.e. more structured) since it seems like it will be easier to expand  if/when we add new symbol values. OTOH, YAGNI and all :)

sbc100 reviewed (approved):
LGTM.  Maybe add "NFC" to the PR title?

wingo commented:
Thanks for the review @binji and @sbc100 :)  AFAICS this can be merged.  (Cheeky request: may I have merging privileges?  Very OK to say no.)

binji commented:
> may I have merging privileges?

Fine with me, I kinda thought you already could as long as there was a reviewer. Lemme check the settings.
